Issue 80 abstract page rendering

enkelmedia opened the pull request:
Hi!

This PR adds a new abstraction for the rendering: `IPageRenderer` with two different implementations:

* **Razor** which is the default and the one used in the current code base.
* **Http** which will perform a HTTP-request to fetch the content.

The `RazorPageRenderer` in registered during ServiceConfiguration but can be overridden if needed. We also use a named HttpClientFactory to allow for custom configuration (see the updated developer docs).

I also made the Render-methods virtual to allow for sub-classes to re-use the logic by adding on top of the functionality by overriding that method.

This PR contains breaking changes in that the ctor for CacheService has both added and removed dependencies. This would only impact someone that inherits this or uses it in some other way. Since we need to add the `IPageRenderer` interface as a dependency for the CacheService there is really no good way to add this without breaking the ctor contract.

Some questions that might impact the way forward:

**Should we wrap the call to `IPageRenderer.Render()` in a try-catch and log errors in the CacheService OR should each implementation be responsible for error handling?**
Currently, the error handling is performed by each handler.

**Do you think that the "Page Renderer" should be an option/setting or can we rely on DI like in the updated docs?**
I think that an option might be better for "discoverability" but the "pure DI approach" is kind of cleaner, since the developer guide is quite short I would say that this approach works is viable, but do you agree?

```csharp
services.AddUnique<IPageRenderer,HttpPageRenderer>();
```

**I did not add any default custom header to the HTTP-request (to identify requests for search indexing) since this can be done by adding a named HttpClient to the HttpClientFactory. Do you think we should add a default header?**
I've added information in the developer guide about how to add headers and cookies for the requests. My opinion in that this is probably sufficient.

skttl commented:
Looks good so far. I just added some profiling, and it performs really well, maybe it should be the default method anyway?

> I did not add any default custom header to the HTTP-request (to identify requests for search indexing) since this can be done by adding a named HttpClient to the HttpClientFactory. Do you think we should add a default header?

I've added a user-agent header to the httpclient (and changed it in the http context for the razor renderer), to use that as identification of the indexing process. (ba6fbc0)

> Should we wrap the call to IPageRenderer.Render() in a try-catch and log errors in the CacheService OR should each implementation be responsible for error handling?

I think it should be a responsibility of the renderer, so how it is now is good.

**How about async?**
I'm not that averse in doing async stuff. Do you think we should go async on everything? I can see you are using GetAsync on the httpclient.

For reference, (h5is) I merged your changes and added mine in https://github.com/skttl/umbraco-fulltextsearch8/tree/pr/n83_issue-80-abstract-page-rendering

enkelmedia commented:
Hi man!

Thanks for looking in to the PR!

If you think that it should be the default I can only agree. The site where I use my custom thing seems to work fine. This site does not have any protected content and for scenarios like that I guess that the implementor would have to find a way to bypass the protection during indexing. Other than that I would vote for HTTP as the default at the end of the day it's just one line of code to switch if someone has issues with this breaking change.

> I've added a user-agent header to the httpclient (and changed it in the http context for the razor renderer), to use that as identification of the indexing process. (https://github.com/skttl/umbraco-fulltextsearch8/commit/ba6fbc0202d8f92505c845f3effe2773143df165)

Agreed that a custom header is a good thing. I was reminded that sometimes when using proxies (think Cloudfront, cloudflare, WAP) etc in front of the website the request needs to have a "real looking" user agent string. I would vote for setting the user agent to something like `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.80 Safari/537.36` by default and also making this configurable and then add another custom header to use to indicate that the page is rendered for indexing. Not super-important for me but if you don't want to fake the user agent I would vote for making the user agent string an option/setting.

> How about async?

I think that it would be a good idea, I just followed the existing approach in the code base by calling async methods in a synchronous way. I haven't looked at how things are initiated further up the stack but if there are async versions of the things that call down to the cache service it would probably be a good idea. However, it sounds to me like a different task? But again, if you are open to some breaking changes maybe this is the right time to introduce async as well so that we don't have to break things again when introducing async.
